features/crc/mbed_crc.h

```diff
@@ -0,0 +1,145 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_CRC_API_H
+#define MBED_CRC_API_H
+
+/** \addtogroup crc */
+/** @{*/
+#include <stdint.h>
+
+typedef uint64_t crc_data_size_t;
+
+/** CRC Polynomial type
+ *
+ * Differnet polynomials supported
+ */
+typedef enum crc_polynomial {
+    CRC_7BIT_SD,
+    CRC_8BIT_CCITT,
+    CRC_16BIT_CCITT,
+    CRC_16BIT_SD,
+    CRC_16BIT_IBM,
+    CRC_32BIT
+} crc_polynomial_t;
+
+/** CRC width
+ *
+ * CRC value can be 8/16/32 bit
+ */
+typedef enum crc_width {
+    CRC_INVALID = 0,
+    CRC_7 = 7,
+    CRC_8 = 8,
+    CRC_16 = 16,
+    CRC_32 = 32,
+} crc_width_t;
+
+#define CRC_GET_DATA_SIZE(x) ((x) <= 8 ? 1 : ((x) <= 16 ? 2 : 4))
+#define CRC_TOP_BIT(x)       ((x) < 8 ? (1u << 7) : (1u << ((x) - 1)))
+#define CRC_DATA_MASK(x)     ((1u << (x)) - 1)
+
+/** CRC object provides CRC generation through hardware/software
+ */
+class CRC
+{
+public:
+    /** Lifetime of CRC object
+     */
+    virtual ~CRC() {};
+
+    /** Initialize CRC module
+     *
+     *  @return  0 on success or a negative error code on failure
+     */
+    virtual int32_t init()
+    {
+        return 0;
+    }
+
+    /** Deinitialize CRC module
+     *
+     *  @return  0 on success or a negative error code on failure
+     */
+    virtual int32_t deinit()
+    {
+        return 0;
+    }
+
+    /** Get CRC polynomial
+     *
+     * Get the current CRC polynomial
+     *
+     * @return  Polynomial value
+     */
+    virtual uint32_t get_polynomial(void) const = 0;
+
+
+    /** Compute CRC for the data input
+     *
+     *  @param  buffer  Data bytes
+     *  @param  size  Size of data
+     *  @param  crc  CRC
+     *  @return  0 on success, negative error code on failure
+     */
+    virtual int32_t compute(void *buffer, crc_data_size_t size, uint32_t *crc) = 0;
+
+    /** Compute partial start, indicate start of partial computation
+     *
+     *  This API should be called before performing any partial computation
+     *  with compute_partial API.
+     *
+     *  @param  crc  Intial CRC value set by the API
+     */
+    virtual void compute_partial_start(uint32_t *crc)
```

Comment:
> Maybe it would be better to mark this function as pure virtual so implementers doesn't skip it.

Implementation of partial CRC computation shouldn;t be forced. I am not sure if all hardware CRC implementations will support that.

Reply:
From a user and implementer perspective, this is confusing: if *A* and *B* inherits from CRC, when *A* is used as an instance of CRC it may not be replaceable by an instance of *B*.  It is not mentioned that these functions may not be implemented. 

This also has severe implications on usability on hardware CRC if the data used to compute the CRC arrive by chunk: the application should either use a software CRC when a chunk is ready to be processed or recompose the whole data to process before the computation of the CRC if hardware computation is used. 

It may not always be possible to store all the data in memory (as an example consider a large file on disk); the current interface forces the use of Software CRC in such case. 

>  I am not sure if all hardware CRC implementations will support that.

@bulislaw Can we have clarity on this point ?